### PR TITLE
Native ESM

### PR DESCRIPTION
**Pull request by smorimoto**
## Context

When building actions in ESM, some incompatibilities were caused by packaging using ncc, which caused runtime errors. So the solution is to switch the source to native ESM and generate code for ESM and CJS at build time.

### tsup

The build system is changed to tsup, and tsc is now used only for type checking. This makes support for both esm and cjs much simpler.

### Vitest

Switch to vitest for faster and easier testing in ESM.

### Biome

This project is not very complex, so formatting and linting are switched to biome. This results in faster, simpler setup and less dependencies.

### `.npmignore`

If the files field is already included in package.json, you should not have to add the `.npmignore` file. As you can see by running dry-run, there are much fewer sources in the package now.

```
~/src/github.com/github/dependency-submission-toolkit native-esm
❯ npm pack --dry-run
npm notice 
npm notice 📦  @github/dependency-submission-toolkit@1.2.9
npm notice === Tarball Contents === 
npm notice 1.1kB  LICENSE           
npm notice 2.8kB  README.md         
npm notice 7.6kB  dist/index.cjs    
npm notice 29.0kB dist/index.cjs.map
npm notice 11.1kB dist/index.d.cts  
npm notice 11.1kB dist/index.d.ts   
npm notice 6.9kB  dist/index.js     
npm notice 29.0kB dist/index.js.map 
npm notice 1.6kB  package.json      
npm notice === Tarball Details === 
npm notice name:          @github/dependency-submission-toolkit         
npm notice version:       1.2.9                                         
npm notice filename:      github-dependency-submission-toolkit-1.2.9.tgz
npm notice package size:  17.1 kB                                       
npm notice unpacked size: 100.2 kB                                      
npm notice shasum:        aa2ad1f5115dd6020565878219052266e8c7350c      
npm notice integrity:     sha512-Iphxp7mGIctTS[...]soLs4oK6rz76Q==      
npm notice total files:   9                                             
npm notice 
github-dependency-submission-toolkit-1.2.9.tgz
```

### Unpacked Size

`1.16 MB` -> `0.1 MB`

### CODE REVIEW

**Comment by smorimoto** [comment]
Ready to go!

**Comment by febuiles** [comment]
@smorimoto thank you very much for this contribution there's, a lot we can benefit from in here.

Can I ask you to split this into different PRs? From going over the code once I can think of a few:

1. Upgrade Node to 20.
2. Add ESM-required changes.
3. Add extra tooling (e.g. Biome).

This can help us review and more quickly address each change. Thanks again.

**Comment by smorimoto** [comment]
@febuiles Thanks a lot for the quick response! Actually, I considered that at first, but it's a bit hard to split things into different PRs except Node.js 20 upgrade.

The reasons for each are:

### tsup

This is required to deal with the NCC outputting code which is not ESM compliant.

### Vitest

This is not strictly required, but you will need to make some configuration changes to Jest to test the ESM code with Jest.

### Biome

This is also not strictly required, but the difference is much smaller than changing ESLint configuration for ESM.

File: tsup.config.ts
Comment (smorimoto) [inline tsup.config.ts:11]: I think Node.js 16 is the lower bound in the current "works but deprecated" sense. You should be able to easily update the lower bound to Node.js 20 by simply changing this in the future.

File: tsconfig.json
Comment (smorimoto) [inline tsconfig.json:1]: Just use the strict recommended TSConfig.

File: example/tsconfig.json
Comment (smorimoto) [inline example/tsconfig.json:1]: Just use the strict recommended TSConfig.

**Comment by smorimoto** [comment]
I tried to reduce the diff as much as possible. This makes reviewing easier (I guess).

File: .github/workflows/dependency-submission-example.yml
Comment (smorimoto) [inline .github/workflows/dependency-submission-example.yml:24]: Using prepare instead of prepack eliminates the need for rebuild. But prepack may be less noisy during the development process.

File: .github/workflows/test.yml
Comment (smorimoto) [inline .github/workflows/test.yml:27]: Same.

File: package.json
Comment (smorimoto) [inline package.json:33]: You can choose prepare or prepack depending on your needs.

File: .github/workflows/publish.yml
Comment (smorimoto) [inline .github/workflows/publish.yml:18]: For small projects, the cache will probably just cause thrashing, which is just a waste of compute resources.

File: example/package.json
Comment (smorimoto) [inline example/package.json:1]: I just removed unnecessary things out and ran sort-package-json to organise package.json order.

File: package.json
Comment (smorimoto) [inline package.json:1]: Ran sort-package-json to organise package.json order.

File: package.json
Comment (smorimoto) [inline package.json:10]: Should be unnecessary.

File: package.json
Comment (smorimoto) [inline package.json:5]: Probably misplaced at the top level?

File: .npmignore
Comment (smorimoto) [inline .npmignore:1]: Conflicts with `files` in `package.json` and is not needed: https://docs.npmjs.com/cli/v10/configuring-npm/package-json#files

**Comment by smorimoto** [comment]
@febuiles I left some comments. I hope these help the review process.

**Comment by febuiles** [comment]
I think all of this makes sense, thanks for your contribution. I went over it and things look good, I'll try to carve up some time to do local testing before merging by next week at the latest.

**Comment by smorimoto** [comment]
@febuiles Sorry for the lot of diff 😅
As a side note, the results of the test run for the example action can also be found here: https://github.com/smorimoto/dependency-submission-toolkit/actions/runs/7502291922

File: .github/workflows/dependency-submission-example.yml
Comment (smorimoto) [inline .github/workflows/dependency-submission-example.yml:19]: The old way doesn't seem to work, so for clarity, use the verbose flag here. That said, the short hand should work if you do the following: `-w example` or `-w=example`, not `-w:example`

File: example/npm-detector.ts
Comment (smorimoto) [inline example/npm-detector.ts:113]: `--prod` is already deprecated

File: example/tsconfig.json
Comment (hmaurer) [inline example/tsconfig.json:10]: TIL, this is neat!

**Review by febuiles — APPROVED** [review]

File: example/tsconfig.json
Comment (smorimoto) [reply example/tsconfig.json:10]: I always recommend using this!

**Comment by febuiles** [comment]
@smorimoto Thank you very much for your contribution. Once https://github.com/github/dependency-submission-toolkit/pull/59 is merged we'll do a new major release that will include these changes.

**Comment by smorimoto** [comment]
@febuiles My pleasure! Thank you so much for this fantastic package! The OCaml community is really actively taking advantage of this 🙂

**Comment by smorimoto** [comment]
Confirmed v2 works with the ESM GHA! Thanks a lot!

**Comment by smorimoto** [comment]
@febuiles @hmaurer Offtopic: In the OCaml community, the OCaml core team has already released a stable version of the dependency submission action, and we want to make it well-integrated with GitHub, but is there a way to do this? I couldn't find that anywhere.

**Comment by febuiles** [comment]
@smorimoto I think @jonjanego could help with that!

**Comment by febuiles** [comment]
@smorimoto the ESM changes are causing at least one issue I'm aware of: https://github.com/advanced-security/maven-dependency-submission-action/actions/runs/7505596002/job/20435167951, did we miss any steps here?

**Comment by smorimoto** [comment]
@febuiles The reason for that seems to be that the `index.ts` does not export `DependencyScope`.

https://github.com/github/dependency-submission-toolkit/blob/86a613fb471daa83163d04299f1143fa97d67627/src/index.ts#L1

How about exposing all public APIs in the package root like this?

```ts
export * from './manifest.js'
export * from './package-cache.js'
export * from './package.js'
export * from './snapshot.js'
```

**Comment by smorimoto** [comment]
@febuiles See this PR! https://github.com/github/dependency-submission-toolkit/pull/63

**Pull request merged**
